ONSdigital/eq-questionnaire-runner
# Add Publisher Backend - Pub/Sub

### What is the context of this PR?
Adds support for GCP's Pub/Sub as a publisher backend for fulfilment requests (IR).

### How to review 
Ensure you can request an IR using Pub/Sub and that the fulfilment request payload is as expected.

For dev, the fulfilment request is outputted to logs. To use GCP Pub/Sub, you need to:

---

1. **__INITIAL SET UP__**

---

**TESTING LOCALLY**: (Using launcher and runner locally with GCP Pub/Sub)
	
 - Build a GCP environment using Concourse/Terraform. (this will create a Pub/Sub topic called `eq-fulfilment-topic`)
 - Authenticate with gcloud SDK using:
	- `gcloud auth login`
	- `gcloud auth application-default login`
	- `gcloud config set project <your-project-id>`
 - Set the following env vars:
	- `EQ_PUBLISHER_BACKEND: pubsub`
	- `EQ_PUB_SUB_TOPIC_ID: eq-fulfilment-topic`
 - Install the latest Pipfile.lock packages.


**TESTING IN GCP**: (Using deployed launcher and runner with GCP Pub/Sub)
  - Build a GCP environment using Concourse/Terraform while ensuring you are using the correct branch and image tag. For example, when deploying with Concourse, be sure to update the Git source **branch** and deployment **image tag** to:
	
	Note: Since we don't push to GCR for runner, you will need to build and push your own image to GCR to deploy to Concourse. Alternatively you can use the image I have already pushed.
	- **runner**: add-pub-sub-publisher
	
	- **launcher**: add-census-claims-to-ir-schema

---

2. __**TEST MESSAGE PUBLICATIONS**__

---
  - Go to Pub/Sub in GCP Console and click the `eq-fulfilment-topic` topic.
  - Click `VIEW MESSAGES`, this will prompt you to create a `Subscription` which is needed to view messages.
  - Request an IR using `test_individual_response` schema. If you are testing locally, you will need to use https://github.com/ONSdigital/eq-questionnaire-launcher/pull/21 (the branch has already been updated in the docker-compose YAML)
  - Click `VIEW MESSAGES` in the GCP UI, select the subscription you just created and click `PULL`. You should now see the IR response fulfilment request payload. Test multiple scenarios and ensure the payload matches the specification.

Known issue:
- The GCP Pub/Sub library relies on `grpc` and threading for which `gevent` only has experimental support for. Further investigation will be done to try to use `gthread` over `gevent`.

### Checklist
